pkg/pillar/hypervisor/kvm.go

```diff
@@ -929,12 +939,180 @@ func getFmlCustomResolution(status *types.DomainStatus, globalConfig *types.Conf
+	}
+
+	return mds
+}
+
+func (p pciDevice) pciLongWOFunction() (string, error) {
+	pciLongSplit := strings.Split(p.pciLong, ".")
+	if len(pciLongSplit) == 0 {
+		return "", fmt.Errorf("could not split %s", p.pciLong)
+	}
+	pciWithoutFunction := strings.Join(pciLongSplit[0:len(pciLongSplit)-1], ".")
+
+	return pciWithoutFunction, nil
+}
+
+type pciAssignmentsTemplateFiller struct {
+	multifunctionsDevices multifunctionDevs
+	file                  io.Writer
+}
+
+func (f *pciAssignmentsTemplateFiller) pciEBridge(pciID int, pciWOFunction string) error {
+	pciChildTemplateVars := struct {
+		PCIId int
+		Bus   int
+	}{
+		PCIId: 0,
+		Bus:   pciID,
+	}
+
+	tPCIeBridge, err := template.New("qemuPCIeBridge").Parse(qemuPCIPassthruBridgeTemplate)
+	if err != nil {
+		return fmt.Errorf("parsing qemuPCIPassthruBridgeTemplate failed: %w", err)
+	}
+	if err := tPCIeBridge.Execute(f.file, pciChildTemplateVars); err != nil {
+		return logError("can't write PCIe bridge Passthrough to config file (%v)", err)
+	}
+	f.multifunctionsDevices[pciWOFunction].bridgeBus = fmt.Sprintf("pcie-bridge.%d", pciID)
+
+	return nil
+}
+
+func (f *pciAssignmentsTemplateFiller) do(file io.Writer, pciAssignments []pciDevice, pciID int) error {
+	if len(pciAssignments) == 0 {
+		return nil
+	}
+
+	pciPTContext := struct {
+		PCIId        int
+		PciShortAddr string
+		Xvga         bool
+		Xopregion    bool
+		Bus          string
+		Addr         string
+	}{PCIId: pciID, PciShortAddr: "", Xvga: false, Xopregion: false}
+
+	tPCI, _ := template.New("qemuPCI").Parse(qemuPciPassthruTemplate)
+	for i, pa := range pciAssignments {
+		pciPTContext.Xopregion = false
+		pciPTContext.Xvga = pa.isVGA()
+		pciPTContext.Bus = fmt.Sprintf("pci.%d", pciPTContext.PCIId)
+		pciPTContext.PciShortAddr = types.PCILongToShort(pa.pciLong)
+
+		if vendor, err := pa.vid(); err == nil {
+			// check for Intel vendor
+			if vendor == "0x8086" {
+				if pciPTContext.Xvga {
+					// we set opregion for Intel vga
+					// https://github.com/qemu/qemu/blob/stable-5.0/docs/igd-assign.txt#L91-L96
+					pciPTContext.Xopregion = true
+				}
+			}
+		}
+
+		// for non-multifunction devices, every pci device gets a "pcie-root-port"
+		if !f.multifunctionsDevices.isMultiFunction(pa) || f.multifunctionsDevices.isFirstMultiFunction(pa) {
+			tRootPortPCI, _ := template.New("qemuRootPortPCI").Parse(qemuRootPortPciPassthruTemplate)
+			if err := tRootPortPCI.Execute(file, pciPTContext); err != nil {
+				return logError("can't write Root Port PCI Passthrough to config file (%v)", err)
+			}
+		}
+		if f.multifunctionsDevices.isMultiFunction(pa) {
+			pciLongWoFunc, err := pa.pciLongWOFunction()
+			if err != nil {
+				logrus.Warnf("retrieving pci address without function failed: %v", err)
+				continue
+			}
+			if f.multifunctionsDevices.isFirstMultiFunction(pa) {
+				f.pciEBridge(pciPTContext.PCIId, pciLongWoFunc)
+			}
+
+			pciPTContext.Bus = f.multifunctionsDevices[pciLongWoFunc].bridgeBus
```

Comment:
I'm not sure it can really happen, but if we first handle a multifunctional device that is not the "first" one, we might misfill a field here, as .bridgeBus is not initialized yet (it's set in the pciEBridg function, which has not been called yet in this case).
Maybe we could simplify the logic by adding a flag for each multifunctional device, like `firstHandled`, and avoid relying on indices?

Reply:
I changed it to use a dedicated map to check instead of relying on the index.